Add type hints

Morxemplum opened the pull request:
This is kind of a part 1 of refactoring the script to make it a ton more maintainable. This merge request refactors the script to use type hints, an optional albeit incredibly useful feature of Python. By adding type hinting into the script and development workflow, we can eliminate a whole class of errors. 

Since discord.py doesn't come with its own type stubs by default, I autogenerated them using my type checker so hopefully all typing logic related to discord.py should be handled. Any remaining errors that I found from my type checker have been explicitly ignored, as we don't want to change discord.py logic.

In addition, I also created a config file specifically to the project so that we can not only configure where type checkers should be looking, but also make it so that we can globally ignore certain issues that can be raised by the pyright type checker. For example, I chose to ignore the `reportUnusedCallResult` rule because I find this rule to be incredibly pedantic (even for typed languages) and will just hurt legibility more than it offers a benefit. 

I am open to blacklisting more rules if you do find them to be too pedantic (i.e. should we allow implicit Any if we want to avoid the usage of `cast`?). I want the code base to still be workable for everyone, but at the same time we should make sure that we can catch and eliminate any typing errors that can pop up when modifying the code.

And if you haven't already, you should consider using a static type checker for Python code. I modified the README to include two type checkers that can be used: the popular [Pylance](https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-pylance), but also the FOSS [basedpyright](https://docs.basedpyright.com/latest/) which is what I'm using, as regular pyright has fallen behind quite a bit in features.

Commit:
Add type hints to script

Bauumm reviewed (changes requested):
I don't really have time to review the details right now. You may need to wait until next week for that.
However I can already tell you a few things:
- remove all the discord.py files. These should not be needed in our project. It is an error with your linter configuration if it doesn't have / find them, not with this repo.
- There is a lot of needless information in your code changes, e.g. `score_text: str = f"...`. The type annotation is only needed if it's not clear what type something is. Anything that is inferable does not need it. Your linter should be able to see that this is a string even if you don't tell it explicitly. (Let alone a human)

Commit:
Fix check_scores_task with new methods for default state

Commit:
Remove explicit type hints if they can be properly inferred
For the sake of legibility, I decided to remove type hints from many different declarations, if my type checker can reasonably infer the correct types. If my type checker cannot resolve the type automatically, or it is too confusing to automatically know what the type is for a variable, a type hint is explicitly shown.

Commit:
Add autogenerated discord.py type stubs to .gitignore

Commit:
Add pyright config file

Commit:
Update README.md

Morxemplum commented:
Alright, following your comments, I did make some changes.

> remove all the discord.py files. These should not be needed in our project. It is an error with your linter configuration if it doesn't have / find them, not with this repo.

Vipre wasn't a fan of this either. The reason why they exist is because the `ext` module from discord.py does not have type stubs associated with it (see attached screenshot). The way to fix this is to use autogenerated type stub files. However, given that these files are autogenerated, I decided to add the `typings/discord` directory to .gitignore so that these will not show up in the repository. I do think the `typings` directory will be of use if we want to push some of the typedef stuff into a separate file to help clean up the script and make it look neater. 

<img width="1102" height="369" alt="Screenshot_20250824_170253" src="https://github.com/user-attachments/assets/0526e0a1-6a17-4ffc-8f3b-8ea676528b29" />

> There is a lot of needless information in your code changes, e.g. score_text: str = f".... The type annotation is only needed if it's not clear what type something is. Anything that is inferable does not need it. Your linter should be able to see that this is a string even if you don't tell it explicitly. (Let alone a human)

I resolve this with commit https://github.com/Open-Hexagon/oh-leaderboard-bot/pull/9/commits/35bb519d4d3f46fb3aabad11a97fa63e5062e6e1. Please feel free to read the commit description to know more about what happened.

Bauumm commented:
Can't the linter get the data (class definitions etc) from the source files of the library? Why do you need to generate extra files?
If this is about linting performance, then that is also not something our project should tackle. This is a discord bot.

Bauumm commented on main.py:
```diff
@@ -2,18 +2,75 @@
 import discord
 from discord.ext import tasks
 
+from collections.abc import Sequence
+from discord import Message
+from discord.channel import TextChannel
+from typing import cast, TypedDict
```
Are these imports really required? As far as I know type annotations have no runtime effect for python, so it seems a bit silly to import something just to add them.
This is just nitpicking though, since there is no real issue with having a few more imports.

Morxemplum replied:
Yes, all of these imports are needed to properly define the type hints and typing that are used in the script. Remove any of these imports and you will immediately run into errors.

The only import that I can possibly think of removing here is the Sequence import, and that is if you want to replace the one spot where it is used with a List

Bauumm commented on main.py:
```diff
@@ -2,18 +2,75 @@
-def rreplace(s, old, new):
+## TYPEDEFS
+
+# Generic JSON typedef that uses type erasure.
+type JSON_Dict = dict[str, object]
```
Is the type `object` somehow different to `Any`?

Morxemplum replied:
Type checkers in their more strict settings will not want you using `Any` (especially basedpyright's "Recommended" mode). The main reason why is because if you use `Any`, then what you are doing is you are essentially disabling the type checking for that part of the code. `object` is a lot more formal here in that you still have a generic type, but you are able to better enforce type checking measures.

But as I have mentioned in my pull request description, if you want to disable rules that are related to banning `Any` (including the use of `cast` to address cases of implicit `Any`), please tell me and I'll gladly blacklist them.

Bauumm replied:
Yeah I think it's a better idea to disable the rules before we spend ages trying to find some nice ways to work around them.

Morxemplum replied:
I completely understand allowing implicit `Any` (as this is definitely one of pyright's more pedantic rules and I am personally not a fan of it either), but are you sure you want to allow explicit `Any`?

Also keep in mind that by allowing implicit `Any`, that will remove our `cast` calls, but those variables will need explicit type hints as the type checker will not be able to properly infer the type on its own.

Bauumm replied:
As I said in the initial review comment. For me at least pythons type checking tools are more of an optional tool to help in places where it's annoying to tell what type something is for a human.

If I wanted everything to be typed I would not use python.

Bauumm replied:
So yes I am fine with implicit `Any`.

Bauumm commented on main.py:
```diff
@@ -35,38 +95,39 @@ async def setup_hook(self) -> None:
         try:
             with open("saved_state.json") as fp:
-                saved_state = json.load(fp)
+                saved_state = cast(Task_State, json.load(fp))
```
I think I understand why your linter wants to have it, but it is very ugly.
The reason people use languages like python is precisely because you don't need to do things like this.
(the same applies to the other points where you did this)

Morxemplum replied:
I answer this in my previous comment: this is to address cases of implicit `Any`. If you would like me to blacklist this rule and remove any `cast` calls, feel free to tell me and I'll do so.

Bauumm commented on main.py:
```diff
@@ -35,38 +95,39 @@ async def setup_hook(self) -> None:
         self.create_lookup_table()
 
         # start the task to run in the background
-        self.check_scores_task.start()
+        self.check_scores_task.start() 
         log(f"Started background task for leaderboard checking.")
 
     @tasks.loop(seconds=REFRESH_TIME)  # task runs every REFRESH_TIME seconds
     async def check_scores_task(self):
-        saved_state = {}
+        saved_state: Task_State
         try:
             with open("saved_state.json") as fp:
-                saved_state = json.load(fp)
+                saved_state = cast(Task_State, json.load(fp))
         except FileNotFoundError:
-            pass  # just uses defaults
-        saved_state["video_queue"] = saved_state.get("video_queue", [])
-        saved_state["last_call_timestamp"] = saved_state.get("last_call_timestamp", 0)
-        saved_state["recent_scores"] = saved_state.get("recent_scores", [])
-
+            # Create a backup state with default values
+            saved_state = {
+                "video_queue": [],
+                "last_call_timestamp": 0,
+                "recent_scores": []
+            }
```
This change will break things if your saved state only has 1 or 2 of the 3 properties (which has been the case after some updates before).
I also don't understand why this change is required. Shouldn't your linter be able to infer the types from the get call?

Morxemplum replied:
The problem with the default `get` approach is there is room for the `Any` type. If you want to allow `Any`, then we could try and revert this back. Otherwise, I'll see if I can find a different approach to fix this regression.

Bauumm replied:
Why is there room for the any type? saved_state has a well defined type, so getting a specific property can only have the type in saved_state or the type of the default value, both of which is fixed here.

Morxemplum replied:
Whoops. I must've misremembered warnings I got earlier when refactoring the code (before I established TypedDicts). But yeah I guess I can reintroduce those three lines in the try block, as that should fix the regression.

Bauumm commented on main.py:
```diff
@@ -22,11 +79,14 @@ def log(s : str):
+
 class leaderboard_client(discord.Client):
-    def __init__(self, *args, **kwargs):
+    def __init__(self, *args: object, **kwargs: object):
```
Here again not sure if `object` is somehow different to `Any` or just not having an annotation. I'd prefer to not have one here. I don't think your linter is going to learn anything from this either.

Morxemplum replied:
Removing the type annotation here will raise warnings from my type checker regarding the rules `reportUnknownParameterType` and `reportMissingParameterType`. It cannot properly infer the type regarding these parameters.

Any parameter for a method that isn't `self` needs to have a type hint.

Bauumm replied:
Starred parameter expressions are inherently of an unkown type. Seems like a bad design decision in the type checker to require an annotation here to me.

Bauumm replied:
But if there is no nice way to change this at all I would be fine with keeping this anyway.

Morxemplum replied:
> Starred parameter expressions are inherently of an unkown _[sic]_ type . Seems like a bad design decision in the type checker to require an annotation here to me.

That's your opinion, but yes starred parameters like that can be very generic. That's why `object` works best here, as it resembles that it could be almost any type.

Bauumm replied:
With implicit any allowed this part could also be omitted now.

Bauumm commented on main.py:
```diff
@@ -35,38 +95,39 @@ async def setup_hook(self) -> None:
-        channel = self.get_output_channel()
+        
+    async def send_wrs(self, scores_json: list[Score_Dict], saved_state: Task_State):
+        channel: TextChannel = self.get_output_channel()
```
The linter can infer this from the return type of the method.

Bauumm replied:
I just noticed, you also added spaces in a blank line here, don't do that, even the most lenient style checkers will yell at you.

Morxemplum replied:
> I just noticed, you also added spaces in a blank line here, don't do that, even the most lenient style checkers will yell at you.

Don't know how those even got there, but alright then.

Bauumm commented on main.py:
```diff
@@ -79,7 +140,7 @@ async def send_wrs(self, scores_json, saved_state):
                 level_options_str = urllib.parse.quote(json.dumps(score["level_options"]))
 
                 try:
-                    lb_scores = requests.get(f"{LB_API_SERVER}/get_leaderboard/{pack_ID_str}/{level_ID_str}/{level_options_str}").json()
+                    lb_scores: JSON_Dict = cast(JSON_Dict, requests.get(f"{LB_API_SERVER}/get_leaderboard/{pack_ID_str}/{level_ID_str}/{level_options_str}").json())
```
Can the linter not tell that the `.json()` function returns a json dict? or is this just because you wanted your own custom `JSON_Dict` type.
Might be easier to get working if you set the type to just `dict`.

Morxemplum replied:
JSON Dicts are a very broad and generic type, and are *very* open with how they can be structured. That's the reason why I used `TypedDict`s. 

With the `TypedDict`, I am able to take specific JSON Dicts and define the types for their attributes so that it resolves dozens of type checking warnings/errors related to retrieving values from them. Removing those `TypedDict`s will have to involve a ton of annoying assert statements and `isinstance` checks to make sure that attributes are the correct type.

Bauumm replied:
> Removing those TypedDicts

That's not what I was suggesting. I was just talking about the JSON_Dict type.

Morxemplum replied:
Oh. This generic typedef is only used here because we don't actually try and fetch any of the values contained within it. This saves us from having to define and use another TypedDict

Bauumm replied:
Oh also I just noticed the type annotation here is actually wrong as well. The api returns a list of dicts, not a dict.
So to fix this you could probably set the type of lb_scores to just `list`. We can have the contents of the list be any/object.

Bauumm commented on main.py:
```diff
@@ -97,32 +158,32 @@ async def send_wrs(self, scores_json, saved_state):
                         level_name = self.pack_lookup[pack_ID]["levels"][level_ID][0]
 
                     num_diffs = self.pack_lookup[pack_ID]["levels"][level_ID][1]
-
                     mult = f"{score['level_options']['difficulty_mult']:.6g}"
+                    diff = ""
 
-                    diff_str = ""
                     if num_diffs > 1:
-                        diff_str = f" [x{mult}]"
+                        diff = f" [x{mult}]"
                     # if level has only 1 difficulty, but score wasn't set on x1, something is wrong
                     elif mult != "1":
                         log(f"WARNING: Level {level_ID} may have added difficulty mults, refreshing  cache.")
                         self.create_lookup_table()
-                        diff_str = f" [x{mult}]"
+                        diff = f" [x{mult}]"
 
                     player = score["user_name"]
-                    run_length = round(score["value"], 3)
+                    run_length: float = round(score["value"], 3)
```
Type annotations in the TypedDict already say this is a float. No need to say it again.

Morxemplum replied:
Missed a spot. Will fix.

Bauumm commented on main.py:
```diff
@@ -97,32 +158,32 @@ async def send_wrs(self, scores_json, saved_state):
                             saved_state["recent_scores"].remove(last_score)
 
                     edited = False
+                    msg: Message | None = None
```
This is only initialized in two places. Both of them use a function that should specify a return type. (At least discord.py docs have it, they both throw if it doesn't get a message)

Morxemplum replied:
This specifies that msg is an optional type of `Message`. I address in a couple of comments later why I do this.

Bauumm commented on main.py:
```diff
@@ -133,7 +194,8 @@ async def send_wrs(self, scores_json, saved_state):
                                 break
-
+                            
+                            assert isinstance(last_score["message_id"], int), "Attempt to append text into a previous score message, but it doesn't exist"
```
All scores in the recent_scores field of the saved_state have a message id. Your types should reflect that to make this assertion not needed.

Morxemplum replied:
If you take a look at the `Score_Dict` typedef, I state that `message_id` is an *optional* type of `int`. The reason why I made it optional is because the `message_id` attribute is not immediately assigned.

This is just insurance to make sure that the `message_id` field was properly assigned a message ID value.

Bauumm replied:
How does this ensure that the field is set?
The field is optional so your type checker isn't going to complain either way and at runtime it'd just have an exception if it somehow failed.
I don't see how what you are doing is helping.
If we look at the type hints from a pure human perspective I think having it not optional would be fine as reference.

Morxemplum replied:
When `message_id` is **not set**, getting a value with that key gives you `None`. That's the entire point of an *optional* attribute. Unlike the other attributes, `message_id` is not set until your code manually assigns it at the end of this method.

When trying to access the value of `message_id`, the type checker thinks that `message_id` either has an assigned value (`int`), or there isn't one (`None`). It doesn't know that it will always be assigned, and it should not be making such assumptions (as that would be incredibly bad for a type checker).

As you say yourself, "All scores in the recent_scores field of the saved_state have a message id." This means that we can make an assert statement that the `message_id` of `last_score` will have a value. This not only will raise an error in the seemingly impossible case it happens, but it also will tell a type checker: "Hey! There's always going to be a value here. It can't be `None`". This ensures you can call `channel.fetch_message` and the type checker won't raise an error, as you cannot call that method with `None`.

Bauumm replied:
There is a much simpler way to tell the type checker "Hey! There's always going to be a value here. It can't be None".
You could just not have the message id be optional. While it may be inconsistent for a short moment at runtime it is largely correct, still serves as reference and fixes type checker issues without needing to add extra code.

Morxemplum replied:
> You could just not have the message id be optional. While it may be inconsistent for a short moment at runtime it is largely correct

That's bad design. We don't want inconsistency in the code. We shouldn't settle for "largely correct", we want to strive for *completely* correct. Cutting corners like this is going to sacrifice the maintainability of the code, as anybody who wants to contribute is gonna have to learn this little quirk with this attribute: "Okay even though we make it a required attribute, it technically could be unassigned and have no value". The code should communicate this from a type checking perspective. I wrote a comment explaining why this attribute is optional.

Bauumm replied:
Since I approved the assert in another place now it'd be inconsistent to not approve it here.

Bauumm commented on main.py:
```diff
@@ -149,15 +211,17 @@ async def send_wrs(self, scores_json, saved_state):
                         saved_state["recent_scores"].append({**score, "message_id": msg.id})
                     
-                    if rank == 1:
+                    if rank == 1 and msg:
                         saved_state["video_queue"].append({**score, "message_id": msg.id})
+                    # Logically speaking, msg should always be bound, but if it ends up being unbound, log it.
+                    elif not msg:
+                        log("Error: A previous message for a score was not found, and no new message was made!")
```
Well if logically speaking it always works, then it's an issue with your tools that they can't tell.
(Even without the if, we'd still get an exception. Which I think is acceptable for something that should logically never happen)

Morxemplum replied:
While reading the code looks like we'll have 100% coverage, doesn't mean that we have 100% coverage. The only way we can confirm that `msg` is always bound is if we have a code coverage tool that can confirm that with all covered cases, `msg` is bound. 

If that ends up being the case, then we can simplify this code and then use an ignore and make a comment stating that we can confirm `msg` will always be bound.

Bauumm replied:
Even then, technically if there really was a way for msg to be None. I'd be fine with it just throwing an exception rather than needing this extra code.

Morxemplum replied:
Okay, if you insist, I'll replace the much more lenient logging with the stricter exception. I'll use an assert to pull this off.

Bauumm replied:
You don't need an assert, accessing an attribute of none (which `msg.id` would do if `msg` was None) will create an exception.

I think this behaviour is good for a case that technically cannot happen.

Morxemplum replied:
I'm not gonna use an ignore just because it's gonna allow for sloppier code. Ignores should be used sparingly, and there needs to be a good reason to use one. So, if you are able to use a code coverage tool to confirm that `msg` will always be bound in all cases, then we'll have a sufficient reason to use one. 

The only ignore that is used throughout this pull request is with a discord.py decorator, and I use the ignore because the decorator is being used correctly, despite what my type checker is saying. 

Still, I highly suggest that we use an assert statement to ensure `msg` is not `None`. It does what you are looking for and it allows us to only use 1 line of code. It's not that big of a deal.

Bauumm replied:
I'm not a fan of it, but sure. Add an assert.

Bauumm commented on main.py:
```diff
@@ -149,15 +211,17 @@ async def send_wrs(self, scores_json, saved_state):
         while len(queue) > 0:
-            score = queue[0]
-            has_better = False
+            score: Score_Dict = queue[0]
```
This should be inferrable from the parameter type annotation.

Morxemplum replied:
Another missed type infer. Will fix.

Bauumm commented on main.py:
```diff
@@ -179,6 +243,7 @@ async def check_videos(self, queue):
                 return
             if response_headers["Content-Type"] == "video/mp4":
                 # exists now, edit message to include link
+                assert isinstance(score["message_id"], int), "Can't attach video to a score with no assigned Discord message"
```
Here again if it's in the queue, it will always have a message id. And even if somehow it managed to not have one. I rather it just throws an exception below than bloat up the code with needless statements like this.

Morxemplum replied:
Ditto.

Bauumm replied:
Since I approved the assert in another place now it'd be inconsistent to not approve it here.

Bauumm commented on main.py:
```diff
@@ -22,11 +79,14 @@ def log(s : str):
+        self.pack_lookup: dict[str, OpenHexagon_Pack] = {}
 
-    async def on_ready(self):
+    async def on_ready(self) -> None:
```
The return type should be inferrable, since there is no return statement.

Morxemplum replied:
Will fix

Bauumm commented on pyrightconfig.json:
```diff
@@ -0,0 +1,18 @@
+  ],
+
+  "exclude": [
+    "**/node_modules",
```
Ok I am now biased against this for the sole reason that it uses node.js.
This is a simple one small python file discord bot. Don't make people install a huge node modules folder to develop it.

Do your type annotations work with other linters?

Morxemplum replied:
Sorry, I took a lot of the config from a sample config file and didn't understand what this meant. This is my fault and I can remove this.

Bauumm commented on main.py:
```diff
@@ -193,32 +258,22 @@ async def check_videos(self, queue):
             else:
                 return
 
-    @check_scores_task.before_loop # type: ignore
+    @check_scores_task.before_loop  # pyright: ignore[reportArgumentType]
     async def before_my_task(self):
         await self.wait_until_ready()  # wait until the bot logs in
     
-    def get_output_channel(self):
+    def get_output_channel(self) -> TextChannel:
         channel = self.get_channel(LB_CHANNEL_ID)
         if not channel:
             log(f"ERROR: Could not find channel <{LB_CHANNEL_ID}>.")
-            return
-        assert isinstance(channel, discord.TextChannel), "You have set your output to a channel that isn't a text channel."
+        assert isinstance(channel, TextChannel), "You have set your output to a channel that isn't a text channel."
         return channel
 
     def create_lookup_table(self):
-        all_packs = requests.get(f"{LB_API_SERVER}/get_packs/1/1000")
-
-        # pack_lookup: dict of dicts
-        # {
-        #     pack_id: {
-        #         "pack_name": str
-        #         "levels": {
-        #             level_id: (str, #difficulties)
-        #         }
-        #     }
-        # }
+        all_packs_json: Sequence[Leaderboard_Pack_Metadata] = cast(Sequence[Leaderboard_Pack_Metadata], requests.get(f"{LB_API_SERVER}/get_packs/1/1000").json())
```
I am not sure about other people, but I personally find the old comment much more informative than most of the type annotations you added.
It doesn't make you look up a sub type when you want to check something. Maybe it's worth exploring other ways to define the TypedDict structure at the top of the file? (I think you can do something like `{key: type}`, but not sure if you can nest it)

Morxemplum replied:
Perhaps we can use docstrings to reintroduce this comment at the typedef?

Bauumm replied:
I think it'd be nicer to define the TypedDict like this to have it all in one place:
```python
# A 1:1 mapping of pack metadata with the leaderboard API
class Leaderboard_Pack_Metadata(TypedDict):
    id: str
    name: str
    levels: list[{
        id: str
        name: str
        options: {
            difficulty_mult: list[float]
        }
    }]
```

Bauumm replied:
And you can do the same in other places where applicable of course.

Oh and this is personal preference, but I like having either camel or snake case, not both in one. So I would change `Leaderboard_Pack_Metadata` to `LeaderboardPackMetadata` (same for the other types of course).

Morxemplum replied:
That is not how type hinting works in Python; it is invalid code. That's why I think it's best we just explain this with a docstring under that class: you'll still be able to flesh out the structure and explain it, but you don't sacrifice the accuracy of the typedef.

> Oh and this is personal preference, but I like having either camel or snake case, not both in one.

Shame. I like Pascal case, but I understand you're trying to be more textbook with PEP 8. We'll use camel case instead.

Bauumm replied:
> it is invalid code

oops that's what happens when you code in a github comment and don't run it.
here's a fixed version:
```python
class LeaderboardPackMetadata(TypedDict):
    id: str
    name: str
    levels: list[{
        "id": str,
        "name": str,
        "options": {
            "difficulty_mult": list[float],
        },
    }]
```

And yes the point of this is to have the definition in one place while still having it as a typedef. I don't want to write the same thing twice.

Morxemplum replied:
<img width="1321" height="324" alt="image" src="https://github.com/user-attachments/assets/18230654-6b8b-4935-8842-9fdbf11dc115" />

Seriously, you cannot do this with type hinting in Python. My type checker immediately errors on this: "Dictionary expression not allowed in type expression".

Bauumm replied:
Ok now finally when enabling experimental features in basedpyright, this syntax does work:
```python
type LeaderboardPackMetadata = TypedDict[{
    "id": str,
    "name": str,
    "levels": list[TypedDict[{
        "id": str,
        "name": str,
        "options": TypedDict[{
            "difficulty_mult": list[float],
        }],
    }]],
}]
```
Edit: I made it more consistent now with the parent type using the same syntax as the nested type.

Bauumm replied:
Alright, after some more testing, it actually turns out that while this code works fine in a typedef it doesn't seem to work in other situations while running the code, which seems quite inconsistent.
So I guess we will keep the class based typed dicts. (You can still change it to camel case though)

Bauumm reviewed (changes requested):
I reviewed it in more detail now. But I will say this still feels like a lot of nothing. I can understand type annotations in some places, but you seem obsessed with them. They are optional in python for a reason. It's so you don't need to bother with them at every point, but just at the few points where they are actually helpful. (e.g. the pack lookup dict, or function parameters)

I am not opposed to merging this if you can adress my concerns though.

Morxemplum commented:
> Can't the linter get the data (class definitions etc) from the source files of the library? Why do you need to generate extra files? If this is about linting performance, then that is also not something our project should tackle. This is a discord bot.

I can do this with the `discord` module, but not `discord.ext`. I dunno what the people at discord.py are doing, but it seems like there's no type stubs there. If you have no type stubs, then a type checker simply can not function with the code that is imported from that module.

Bauumm commented:
> If you have no type stubs, then a type checker simply can not function with the code that is imported from that module.

It shouldn't error though. It should just see that none exist and either try to infer things from the source code of the library or assume that things are correct. Similar to how any javascript code is technically valid typescript code.

Morxemplum commented:
> It should just see that none exist and either try to infer things from the source code of the library or assume that things are correct.

Type checkers can only do so many assumptions before it run into problems with implementation. In fact, it does try to "infer things", by autogenerating type stubs if none are available. 

> Similar to how any javascript code is technically valid typescript code.

That's not how it works. As you've said before, Python is a dynamically typed language by default, which means that you are able to get away with certain behavior that you otherwise can't in a statically typed language. Type hinting and checking is not a superset of Python, it's actually more of a subset. It does not permit some of those practices you see in a dynamically typed language -- usually because it tends to be sloppy and can lead to unintended behavior. This is fine if you're just writing a quick and dirty script, but if you want a maintainable project, it's best to use the type hinting and static type checking to eliminate a whole class of errors related to dynamic typing.

Bauumm commented:
> it's actually more of a subset

That depends on your configuration.

> It does not permit some of those practices you see in a dynamically typed language

I may want to use them in some places though. There is a reason the bot isn't written in rust.

> but if you want a maintainable project, it's best to use the type hinting and static type checking to eliminate a whole class of errors related to dynamic typing.

Sure you can eliminate errors relating to dynamic typing by eliminating dynamic typing. But I am fine with using dynamic typing and possibly having those errors in some places while properly type checking in others. That still offers some of the advantages without actually giving up the possibilities of dynamic typing.

Morxemplum commented:
> That depends on your configuration.

Take a following look at the snippet of Python code below

```py
my_value = 1234
print(my_value)
my_value = "String"
print(my_value)
```

This is legal Python code. However, even a lenient type checker will raise an error with this snippet of code. You are assigning a variable that was previously an integer with a string value, changing the type and making it dynamic. The only way to make it work is if you use the `Any` type, but again are you sure you want to allow explicit `Any`?

Bauumm commented:
If I wanted the type checker to warn me I would have written it like this:
```python
my_value: int = 1234
print(my_value)
my_value = "String"
print(my_value)
```
So yes I am sure I want to allow `Any`.
